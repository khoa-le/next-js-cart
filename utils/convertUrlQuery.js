export const convertUrlQuery = (urlPath) => {
  return [
    ...new URLSearchParams((urlPath || "").split(/\?/)[1]).entries(),
  ].reduce((q, [k, v]) => Object.assign(q, { [k]: v }), {});
};
