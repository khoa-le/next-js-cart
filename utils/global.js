export const GlobalConfiguration = {
  ProductListing : {
    PerPage: 12,
  }
};
