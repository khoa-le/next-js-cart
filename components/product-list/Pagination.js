import Link from "next/link";
import { useRouter } from "next/router";
import styled from "styled-components";
import { GlobalConfiguration } from "../../utils/global";

const Pagination = ({ numberProducts }) => {
  const router = useRouter();
  let {
    query: { page },
  } = router;

  page = page == undefined ? 1 : parseInt(page);

  let baseUrl = router.asPath.replace(`page=${page}`, "");
  baseUrl += baseUrl.indexOf("?") > 0 ? "" : "?";
  const nextUrl = baseUrl + "page=" + (page + 1);
  const prevUrl = baseUrl + "page=" + (page - 1);
  return (
    <StyledPagination>
      <nav aria-label="Page navigation">
        <ul className="pagination mb-0">
          {page > 1 ? (
            <li className="page-item ml-0">
              <Link href={prevUrl}>
                <a className="page-link previous" aria-label="Previous">
                  <span aria-hidden="true">« Previous</span>
                </a>
              </Link>
            </li>
          ) : null}

          {numberProducts == GlobalConfiguration.ProductListing.PerPage+1 ? (
            <li className="page-item">
              <Link href={nextUrl}>
                <a className="page-link next" aria-label="Next">
                  <span aria-hidden="true">Next »</span>
                </a>
              </Link>
            </li>
          ) : null}
        </ul>
      </nav>
    </StyledPagination>
  );
};

const StyledPagination = styled.div`
  text-align: center;
`;
export default Pagination;
