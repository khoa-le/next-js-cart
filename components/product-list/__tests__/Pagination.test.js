import { render, cleanup,findByText } from "@testing-library/react";
import Pagination from "../Pagination";
import * as nextRouter from "next/router";
import { GlobalConfiguration } from "../../../utils/global";

nextRouter.useRouter = jest.fn();


afterEach(cleanup);

describe("Pagination", () => {
  test("Can render next pagination", () => {
    nextRouter.useRouter.mockImplementation(() => ({
      query: { page: 1 },
      asPath: "/products",
    }));
    const { container, getByText,findByDisplayValue } = render(
      <Pagination
        numberProducts={GlobalConfiguration.ProductListing.PerPage + 1}
      />
    );
    const nextLinkElement = getByText("Next »");
    expect(container.firstChild).toMatchSnapshot();
    expect(nextLinkElement).toBeInTheDocument();
    expect(container.querySelector("ul.pagination li .previous")).toBeNull()
  });
  test("Can render prev and next pagination", () => {
    nextRouter.useRouter.mockImplementation(() => ({
      query: { page:2 },
      asPath: "/products",
    }));
    const { container, getByText } = render(
      <Pagination
        numberProducts={GlobalConfiguration.ProductListing.PerPage + 1}
      />
    );
    const prevLinkElement = getByText("« Previous");
    const nextLinkElement = getByText("Next »");

    expect(container.firstChild).toMatchSnapshot();
    expect(prevLinkElement).toBeInTheDocument();
    expect(nextLinkElement).toBeInTheDocument();
  });

  test("Can render prev button and hide next when number product < perPage +1",()=>{
    const {container, getByText} = render(
      <Pagination numberProducts={GlobalConfiguration.ProductListing.PerPage-1}/>
    )

    const prevLinkElement = getByText("« Previous");
    expect(container.firstChild).toMatchSnapshot();
    expect(prevLinkElement).toBeInTheDocument();
    expect(container.querySelector("ul.pagination li .next")).toBeNull()

  })

  
});
