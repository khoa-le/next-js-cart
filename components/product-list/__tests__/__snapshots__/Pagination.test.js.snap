// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Can render next pagination 1`] = `
<div
  class="sc-AxjAm jWlVxK"
>
  <nav
    aria-label="Page navigation"
  >
    <ul
      class="pagination mb-0"
    >
      <li
        class="page-item"
      >
        <a
          aria-label="Next"
          class="page-link next"
          href="/products?page=2"
        >
          <span
            aria-hidden="true"
          >
            Next »
          </span>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`Pagination Can render prev and next pagination 1`] = `
<div
  class="sc-AxjAm jWlVxK"
>
  <nav
    aria-label="Page navigation"
  >
    <ul
      class="pagination mb-0"
    >
      <li
        class="page-item ml-0"
      >
        <a
          aria-label="Previous"
          class="page-link previous"
          href="/products?page=1"
        >
          <span
            aria-hidden="true"
          >
            « Previous
          </span>
        </a>
      </li>
      <li
        class="page-item"
      >
        <a
          aria-label="Next"
          class="page-link next"
          href="/products?page=3"
        >
          <span
            aria-hidden="true"
          >
            Next »
          </span>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`Pagination Can render prev button and hide next when number product < perPage +1 1`] = `
<div
  class="sc-AxjAm jWlVxK"
>
  <nav
    aria-label="Page navigation"
  >
    <ul
      class="pagination mb-0"
    >
      <li
        class="page-item ml-0"
      >
        <a
          aria-label="Previous"
          class="page-link previous"
          href="/products?page=1"
        >
          <span
            aria-hidden="true"
          >
            « Previous
          </span>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
